jax/_src/random.py

```diff
@@ -1548,12 +1549,44 @@ def _gumbel(key, shape, dtype, mode) -> Array:
         _uniform(key, shape, dtype, minval=jnp.finfo(dtype).tiny, maxval=1.)))
 
 
-def categorical(key: ArrayLike,
-                logits: RealArray,
-                axis: int = -1,
-                shape: Shape | None = None) -> Array:
+def _top_k(operand, k, axis):
+  # TODO: Remove once jax/issues/25867 is fixed.
+  operand = jnp.moveaxis(operand, axis, -1)
+  values, indices = lax.top_k(operand, k)
+  values = jnp.moveaxis(values, -1, axis)
+  indices = jnp.moveaxis(indices, -1, axis)
+  return values, indices
+
+
+def _categorical_without_replacement(
+  key: ArrayLike,
+  logits: RealArray,
```

Comment:
`key` and `logits` can be `Array` here I think, since they are cast to array by the calling function. Then you should be able to use the `.shape` attribute below.